bionic/provider.py

```diff
@@ -843,6 +863,131 @@ def wrapped_compute_func(dep_values):
         return outer_tasks
 
 
+class NewOutputDescriptorProvider(WrappingProvider):
```

Comment:
Missed this comment the first time. I feel like `OutputDescriptorProvider` is missing something, since every provider has an output descriptor -- what's distinctive about this one is that it replaces the output descriptor of another provider.

I agree that these providers should have docstrings though.